dove/src/transaction.rs

```diff
@@ -0,0 +1,875 @@
+                    values.push(ScriptArg::U8(
+                        arg.parse()
+                            .map_err(|err| parse_err(name, tp, args_index, arg, err))?,
+                    ));
+                    Ok((signers, args_index + 1, values))
+                }
+                "u64" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    values.push(ScriptArg::U64(
+                        arg.parse()
+                            .map_err(|err| parse_err(name, tp, args_index, arg, err))?,
+                    ));
+                    Ok((signers, args_index + 1, values))
+                }
+                "u128" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    values.push(ScriptArg::U128(
+                        arg.parse()
+                            .map_err(|err| parse_err(name, tp, args_index, arg, err))?,
+                    ));
+                    Ok((signers, args_index + 1, values))
+                }
+                "address" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    values.push(ScriptArg::Address(Address::from_str(arg)?.addr));
+                    Ok((signers, args_index + 1, values))
+                }
+                "vector<u8>" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    let buffer = if arg.contains('[') {
+                        parse_vec(arg, "u8")?
+                    } else {
+                        hex::decode(arg)?
+                    };
+                    values.push(ScriptArg::VectorU8(buffer));
+                    Ok((signers, args_index + 1, values))
+                }
+                "vector<u64>" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    values.push(ScriptArg::VectorU64(parse_vec(arg, "u64")?));
+                    Ok((signers, args_index + 1, values))
+                }
+                "vector<u128>" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    values.push(ScriptArg::VectorU128(parse_vec(arg, "u128")?));
+                    Ok((signers, args_index + 1, values))
+                }
+                "vector<address>" => {
+                    let arg = self.argument(args_index, total_args)?;
+                    let address = parse_vec::<Address>(arg, "vector<address>")?
+                        .iter()
+                        .map(|addr| addr.addr)
+                        .collect();
+                    values.push(ScriptArg::VectorAddress(address));
+                    Ok((signers, args_index + 1, values))
+                }
+                &_ => Err(anyhow!("Unexpected script parameter: {}", tp)),
+            },
+        )
+    }
+
+    fn argument(&self, index: usize, total_expected: usize) -> Result<&String, Error> {
```

Comment:
Can't we use `self.args.len()` instead of passing `total_expected`?

Reply:
Also I don't think you need to return `&String` here: always prefer `&[]` over `&Vec<>`, `&str` over `&String` and `&Path` over `&PathBuf`